src/dynamic_honey_badger/dynamic_honey_badger.rs

```diff
@@ -108,6 +215,154 @@ where
+            }
+        }
+        Ok(step)
+    }
+
+    /// Handles an epoch start announcement from a remote `DynamicHoneyBadger`.
+    fn handle_dynamic_epoch_started(
+        &mut self,
+        sender_id: &N,
+        epoch: DynamicEpoch,
+    ) -> Result<Step<C, N>> {
+        debug!(
+            "{:?}@{} received epoch {:?}@{:?}",
+            self.netinfo.our_id(),
+            self.start_epoch,
+            sender_id,
+            epoch,
+        );
+        let mut obsolete = false;
+        self.remote_epochs
+            .entry(sender_id.clone())
+            .and_modify(|e| {
+                if *e < epoch {
+                    // Increment the epoch.
+                    *e = epoch;
+                } else if *e >= epoch {
+                    obsolete = true;
+                }
+            }).or_insert(epoch);
+        if obsolete {
+            return Err(ErrorKind::ObsoleteDynamicEpochStarted.into());
+        }
+        let step = self.process_new_epoch(sender_id, epoch);
+        self.remove_messages_upto(sender_id, epoch);
+        step
+    }
+
+    /// Processes an announcement of a new epoch update received from a remote node.
+    fn process_new_epoch(&mut self, sender_id: &N, epoch: DynamicEpoch) -> Result<Step<C, N>> {
+        // Send any DHB messages for `epoch.start_epoch`.
+        let mut ready_messages = self
+            .outgoing_queue_dhb
+            .remove(&(sender_id.clone(), epoch.start_epoch))
+            .unwrap_or_else(|| vec![]);
+        // Send any HB messages for `epoch`.
+        ready_messages.extend(
+            self.outgoing_queue_hb
+                .remove(&(sender_id.clone(), epoch))
+                .unwrap_or_else(|| vec![]),
+        );
+        let mut step = Step::from(
+            ready_messages
+                .into_iter()
+                .map(|msg| Target::Node(sender_id.clone()).message(msg)),
+        );
+        if epoch.start_epoch == self.start_epoch {
+            // Forward `hb_epoch` to `HoneyBadger` if the `DynamicHoneyBadger` start epochs
+            // match. Any queued messages are stored in the `HoneyBadger` instance.
+            let hb_step = self
+                .honey_badger
+                .handle_message(sender_id, HbMessage::EpochStarted(epoch.hb_epoch))
+                .map_err(ErrorKind::DynamicEpochStarted)?;
+            step.extend(self.process_output(hb_step)?);
+        }
+        Ok(step)
+    }
+
+    /// Removes all messages queued for the remote node from epochs upto `epoch`.
+    fn remove_messages_upto(&mut self, sender_id: &N, epoch: DynamicEpoch) {
+        let earlier_keys_hb: Vec<_> = self
+            .outgoing_queue_hb
+            .keys()
+            .cloned()
+            .filter(|(id, e)| {
+                id == sender_id
+                    && (e.start_epoch < epoch.start_epoch
+                        || (e.start_epoch == epoch.start_epoch && e.hb_epoch <= epoch.hb_epoch))
```

Comment:
Shouldn't this be a strict inequality?
I guess it doesn't matter because the method is called right after the messages with _equal_ epoch have all been sent out, but it still seems safer?